File: ZLivePhoto.Android/app/src/main/java/com/zsz/zlivephoto/core/formats/VivoPlugin.kt
package com.zsz.zlivephoto.core.formats

import com.zsz.zlivephoto.core.BinaryUtils
import com.zsz.zlivephoto.core.FooterUtil
import com.zsz.zlivephoto.core.JpegUtil
import com.zsz.zlivephoto.core.LivePhotoAsset
import com.zsz.zlivephoto.core.Mp4Util
import com.zsz.zlivephoto.core.XmpTemplate
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import kotlin.math.round

/**
 * vivo 双文件动态照片格式（IMG_xxx.jpg + IMG_xxx.mp4）。
 */
internal class VivoPlugin : FormatPlugin() {
    override val name: String = "vivo"
    override val display: String = "vivo 动态照片（JPG + MP4 双文件）"

    private val vivoVersion: Int = 2107

    private fun siblingMp4(path: String): String? {
        val file = File(path)
        val parent = file.parentFile
        val mp4 = if (parent != null) File(parent, "${file.nameWithoutExtension}.mp4").path
                  else "${file.nameWithoutExtension}.mp4"
        return if (File(mp4).exists()) mp4 else null
    }

    private fun buildJson(imageTime: Long, liveId: String): ByteArray =
        FooterUtil.buildFooterJson(linkedMapOf(
            "com.android.camera.imageTime" to imageTime,
            "com.android.camera.livephoto" to liveId,
            "version" to vivoVersion
        ))

    private fun buildUuidBox(jsonBytes: ByteArray, liveId: String): ByteArray {
        val footer = FooterUtil.buildFooter(jsonBytes, liveId, FooterUtil.vivoPrefix)
        val payload = ByteArray(Mp4Util.vivoUuid.size + footer.size)
        System.arraycopy(Mp4Util.vivoUuid, 0, payload, 0, Mp4Util.vivoUuid.size)
        System.arraycopy(footer, 0, payload, Mp4Util.vivoUuid.size, footer.size)

        val result = ByteArray(8 + payload.size)
        BinaryUtils.writeU32BE(result, 0, (payload.size + 8).toLong())
        // "uuid" = 0x75 0x75 0x69 0x64
        result[4] = 0x75.toByte()
        result[5] = 0x75.toByte()
        result[6] = 0x69.toByte()
        result[7] = 0x64.toByte()
        System.arraycopy(payload, 0, result, 8, payload.size)
        return result
    }

    override fun detect(path: String): Int {
        if (!path.endsWith(".jpg", ignoreCase = true) && !path.endsWith(".jpeg", ignoreCase = true))
            return 0

        val xmp = GooglePlugin.sniffXmp(path)
        val info = XmpTemplate.parseMotionXmp(xmp)
        if (info.isMotion) return 0 // 内嵌式不归 vivo 管

        return try {
            val f = File(path)
            val size = f.length()
            FileInputStream(path).use { fs ->
                fs.skip(maxOf(0L, size - 8192))
                val tail = ByteArray(minOf(8192L, size).toInt())
                var read = 0
                while (read < tail.size) {
                    val n = fs.read(tail, read, tail.size - read)
                    if (n < 0) break
                    read += n
                }
                val footer = FooterUtil.parseFooter(tail.copyOfRange(0, read))
                if (footer?.livephotoId == null) return 0
                if (siblingMp4(path) != null) 90 else 40
            }
        } catch (e: Exception) {
            0
        }
    }

    override fun read(path: String, log: (String, String, String) -> Unit): LivePhotoAsset {
        log("info", "按 vivo 双文件动态照片解析", "vivo")
        val data = readBytes(path)
        val footer = FooterUtil.parseFooter(data)
        if (footer?.livephotoId == null)
            throw IOException("JPG 尾部未找到 vivo livephoto 标记")

        val liveId = footer.livephotoId!!
        val mp4Path = siblingMp4(path)
            ?: throw IOException("缺少伴生视频文件：${File(path).nameWithoutExtension}.mp4")

        // JPG 主体（去除 footer）→ 拆 Primary / GainMap
        val body = data.copyOfRange(0, footer.footerStart)
        val (jpegs, _) = JpegUtil.splitJpegs(body)
        if (jpegs.isEmpty()) throw IOException("JPG 主体解析失败")
        val primary = jpegs[0]
        val gainmap = if (jpegs.size > 1) jpegs[1] else null

        // MP4：剥离末尾 vivo uuid box
        val mp4Raw = readBytes(mp4Path)
        val mp4Footer = FooterUtil.parseFooter(mp4Raw)
        var imageTime: Long? = footer.imageTime
        if (mp4Footer != null) {
            if (imageTime == null) imageTime = mp4Footer.imageTime
            val mp4Id = mp4Footer.livephotoId
            if (mp4Id != null && mp4Id != liveId) {
                log("warning", "JPG 与 MP4 的 livephoto ID 不一致：$liveId / $mp4Id", "vivo")
            }
        }

        val video = Mp4Util.stripVivoUuid(mp4Raw)
        if (!Mp4Util.hasFtyp(video))
            throw IOException("伴生 MP4 无效（缺少 ftyp box）")

        val asset = LivePhotoAsset(
            primaryJpeg = primary,
            gainmapJpeg = gainmap,
            videoMp4 = video,
            sourceFormat = name,
        )
        asset.livephotoId = liveId
        asset.imageTime = imageTime
        asset.videoInfo = Mp4Util.getTrackInfo(video) ?: mutableMapOf()

        // 由 imageTime（帧序号）反推封面时间戳
        val imageTimeVal = asset.imageTime
        val fps = asset.videoInfo["fps"] as? Double
        if (imageTimeVal != null && fps != null && fps > 0.0) {
            asset.presentationTsUs = round(imageTimeVal.toDouble() / fps * 1_000_000.0).toLong()
        }
        return asset
    }

    override fun write(
        asset: LivePhotoAsset, outDir: String, stem: String,
        log: (String, String, String) -> Unit, options: MutableMap<String, Any?>
    ): MutableList<String> {
        val liveId = asset.livephotoId ?: FooterUtil.generateLivephotoId()
        val imageTime = asset.effectiveImageTime()

        // JPG：vivo XMP（无 motion 标签）+ vivo footer
        val xmp = XmpTemplate.buildVivoXmp(asset.gainmapLength)
        val primary = JpegUtil.replaceOrInsertXmp(asset.primaryJpeg, xmp)
        val jpgJson = buildJson(imageTime, liveId)
        val jpgFooter = FooterUtil.buildFooter(jpgJson, liveId, FooterUtil.vivoPrefix)
        val gainmapLen = asset.gainmapJpeg?.size ?: 0
        val jpgOut = ByteArray(primary.size + gainmapLen + jpgFooter.size)
        var pos = 0
        System.arraycopy(primary, 0, jpgOut, pos, primary.size); pos += primary.size
        asset.gainmapJpeg?.let {
            System.arraycopy(it, 0, jpgOut, pos, it.size); pos += it.size
        }
        System.arraycopy(jpgFooter, 0, jpgOut, pos, jpgFooter.size)

        // MP4：纯视频流 + 末尾 uuid box
        val video = Mp4Util.stripVivoUuid(asset.videoMp4)
        val mp4Json = buildJson(imageTime, liveId)
        val uuidBox = buildUuidBox(mp4Json, liveId)
        val mp4Out = ByteArray(video.size + uuidBox.size)
        System.arraycopy(video, 0, mp4Out, 0, video.size)
        System.arraycopy(uuidBox, 0, mp4Out, video.size, uuidBox.size)

        val jpgPath = File(outDir, "$stem.jpg").path
        val mp4Path = File(outDir, "$stem.mp4").path
        writeBytes(jpgPath, jpgOut)
        writeBytes(mp4Path, mp4Out)
        log("info", "写出 vivo 格式：$stem.jpg + $stem.mp4（livephoto ID: $liveId）", "vivo")
        return mutableListOf(jpgPath, mp4Path)
    }
}

File: ZLivePhoto.Core/Formats/VivoPlugin.cs
using ZLivePhoto.Core.Models;

namespace ZLivePhoto.Core.Formats;

/// <summary>
/// vivo 双文件动态照片格式（IMG_xxx.jpg + IMG_xxx.mp4）。
/// </summary>
public sealed class VivoPlugin : FormatPlugin
{
    public override string Name => "vivo";
    public override string Display => "vivo 动态照片（JPG + MP4 双文件）";

    private const int VivoVersion = 2107;

    private static string? SiblingMp4(string path)
    {
        var mp4 = Path.Combine(Path.GetDirectoryName(path)!,
            Path.GetFileNameWithoutExtension(path) + ".mp4");
        return File.Exists(mp4) ? mp4 : null;
    }

    private static byte[] BuildJson(long imageTime, string liveId)
    {
        return FooterUtil.BuildFooterJson(new Dictionary<string, object>
        {
            ["com.android.camera.imageTime"] = imageTime,
            ["com.android.camera.livephoto"] = liveId,
            ["version"] = VivoVersion
        });
    }

    private static byte[] BuildUuidBox(byte[] jsonBytes, string liveId)
    {
        var payload = new byte[Mp4Util.VivoUuid.Length +
            FooterUtil.BuildFooter(jsonBytes, liveId, FooterUtil.VivoPrefix).Length];
        Mp4Util.VivoUuid.CopyTo(payload, 0);
        FooterUtil.BuildFooter(jsonBytes, liveId, FooterUtil.VivoPrefix)
            .CopyTo(payload, Mp4Util.VivoUuid.Length);

        var result = new byte[8 + payload.Length];
        System.Buffers.Binary.BinaryPrimitives.WriteUInt32BigEndian(
            result.AsSpan(0, 4), (uint)(payload.Length + 8));
        "uuid"u8.CopyTo(result.AsSpan(4));
        payload.CopyTo(result, 8);
        return result;
    }

    public override int Detect(string path)
    {
        if (!path.EndsWith(".jpg", StringComparison.OrdinalIgnoreCase) &&
            !path.EndsWith(".jpeg", StringComparison.OrdinalIgnoreCase))
            return 0;

        var xmp = GooglePlugin.SniffXmp(path);
        var info = XmpTemplate.ParseMotionXmp(xmp);
        if (info.IsMotion)
            return 0; // 内嵌式不归 vivo 管

        try
        {
            var size = new FileInfo(path).Length;
            using var fs = new FileStream(path, FileMode.Open, FileAccess.Read);
            fs.Seek(Math.Max(0, size - 8192), SeekOrigin.Begin);
            var tail = new byte[Math.Min(8192, size)];
            int read = fs.Read(tail, 0, tail.Length);
            var footer = FooterUtil.ParseFooter(tail.AsSpan(0, read));
            if (footer?.LivephotoId is null)
                return 0;
            return SiblingMp4(path) is not null ? 90 : 40;
        }
        catch
        {
            return 0;
        }
    }

    public override LivePhotoAsset Read(string path, Action<string, string, string> log)
    {
        log("info", "按 vivo 双文件动态照片解析", "vivo");
        var data = ReadBytes(path);
        var footer = FooterUtil.ParseFooter(data);
        if (footer?.LivephotoId is null)
            throw new InvalidDataException("JPG 尾部未找到 vivo livephoto 标记");

        string liveId = footer.LivephotoId;
        var mp4Path = SiblingMp4(path)
            ?? throw new InvalidDataException($"缺少伴生视频文件：{Path.GetFileNameWithoutExtension(path)}.mp4");

        // JPG 主体（去除 footer）→ 拆 Primary / GainMap
        var body = data[..footer.FooterStart];
        var (jpegs, _) = JpegUtil.SplitJpegs(body);
        if (jpegs.Count == 0)
            throw new InvalidDataException("JPG 主体解析失败");
        var primary = jpegs[0];
        var gainmap = jpegs.Count > 1 ? jpegs[1] : null;

        // MP4：剥离末尾 vivo uuid box
        var mp4Raw = ReadBytes(mp4Path);
        var mp4Footer = FooterUtil.ParseFooter(mp4Raw);
        long? imageTime = footer.ImageTime;
        if (mp4Footer is not null)
        {
            imageTime ??= mp4Footer.ImageTime;
            var mp4Id = mp4Footer.LivephotoId;
            if (mp4Id is not null && mp4Id != liveId)
                log("warning", $"JPG 与 MP4 的 livephoto ID 不一致：{liveId} / {mp4Id}", "vivo");
        }

        var video = Mp4Util.StripVivoUuid(mp4Raw);
        if (!Mp4Util.HasFtyp(video))
            throw new InvalidDataException("伴生 MP4 无效（缺少 ftyp box）");

        var asset = new LivePhotoAsset
        {
            PrimaryJpeg = primary,
            GainmapJpeg = gainmap,
            VideoMp4 = video,
            SourceFormat = Name,
            LivephotoId = liveId,
            ImageTime = imageTime
        };
        asset.VideoInfo = Mp4Util.GetTrackInfo(video) ?? new Dictionary<string, object>();

        // 由 imageTime（帧序号）反推封面时间戳
        if (asset.ImageTime.HasValue &&
            asset.VideoInfo.GetValueOrDefault("fps", 0.0) is double fps and > 0)
        {
            asset.PresentationTsUs = (long)Math.Round(asset.ImageTime.Value / fps * 1_000_000);
        }
        return asset;
    }

    public override List<string> Write(LivePhotoAsset asset, string outDir, string stem,
        Action<string, string, string> log, Dictionary<string, object> options)
    {
        string liveId = asset.LivephotoId ?? FooterUtil.GenerateLivephotoId();
        long imageTime = asset.EffectiveImageTime();

        // JPG：vivo XMP（无 motion 标签）+ vivo footer
        var xmp = XmpTemplate.BuildVivoXmp(asset.GainmapLength);
        var primary = JpegUtil.ReplaceOrInsertXmp(asset.PrimaryJpeg, xmp);
        var jpgJson = BuildJson(imageTime, liveId);
        var jpgOut = new byte[primary.Length + (asset.GainmapJpeg?.Length ?? 0) +
            FooterUtil.BuildFooter(jpgJson, liveId, FooterUtil.VivoPrefix).Length];
        int pos = 0;
        primary.CopyTo(jpgOut, pos); pos += primary.Length;
        asset.GainmapJpeg?.CopyTo(jpgOut, pos); pos += asset.GainmapJpeg?.Length ?? 0;
        FooterUtil.BuildFooter(jpgJson, liveId, FooterUtil.VivoPrefix).CopyTo(jpgOut, pos);

        // MP4：纯视频流 + 末尾 uuid box
        var video = Mp4Util.StripVivoUuid(asset.VideoMp4);
        var mp4Json = BuildJson(imageTime, liveId);
        var uuidBox = BuildUuidBox(mp4Json, liveId);
        var mp4Out = new byte[video.Length + uuidBox.Length];
        video.CopyTo(mp4Out, 0);
        uuidBox.CopyTo(mp4Out, video.Length);

        string jpgPath = Path.Combine(outDir, stem + ".jpg");
        string mp4Path = Path.Combine(outDir, stem + ".mp4");
        WriteBytes(jpgPath, jpgOut);
        WriteBytes(mp4Path, mp4Out);
        log("info", $"写出 vivo 格式：{stem}.jpg + {stem}.mp4（livephoto ID: {liveId}）", "vivo");
        return [jpgPath, mp4Path];
    }
}
